proposals/0016-c-interface-compiler-library.md

```diff
@@ -50,7 +50,495 @@ older DXC-based solutions to adopt clang as the preferred HLSL compiler.
+
+Provides support for:
+* Compiling a single entry point to the target shader model
+* Compiling a library to a library target (using -R lib_*)
+* Compiling a rootsignature (-T rootsig_*)
+* Preprocessing HLSL source (-P)
+* Disassembling DXIL container or bitcode
+
+```c++
+struct IDxcCompiler3 : public IUnknown {
+  HRESULT Compile(
+    const DxcBuffer *pSource,            // source text to compile
+    LPCWSTR *pArguments,                 // array of pointers to arguments
+    UINT32 argCount,                     // number of arguments
+    IDxcIncludeHandler *pIncludeHandler, // user-provided interface to handle
+                                         // #include directives (optional)
+    REFIID riid, LPVOID *ppResult);      // IDxcResult: status, buffer, and
+                                         // errors
+
+  HRESULT Disassemble(
+    const DxcBuffer *pObject,            // dxil container or bitcode
+    REFIID riid, LPVOID *ppResult);      // IDxcResult: status, disassembly
+                                         // text, and errors
+};
+```
+#### Shader Linker
+Links a shader and produces a shader blob that can be consumed by the D3D
+runtime.
+
+TODO: Does IDxcCompiler3 cover this already?
+
+```c++
+struct IDxcLinker : public IUnknown {
+public:
+  // Register a library with name to reference it later.
+  HRESULT RegisterLibrary(
+    LPCWSTR pLibName,
+    IDxcBlob *pLib);
+
+  // Links the shader and produces a shader blob that the Direct3D runtime can
+  // use.
+  HRESULT Link(
+    LPCWSTR pEntryName,              // Entry point name
+    LPCWSTR pTargetProfile,          // shader profile to link
+    const LPCWSTR *pLibNames,        // Array of library names to link
+    UINT32 libCount,                 // Number of libraries to link
+    const LPCWSTR *pArguments,       // Array of pointers to arguments
+    UINT32 argCount,                 // Number of arguments
+    IDxcOperationResult **ppResult); // output status, buffer, and errors
+};
+```
+
+#### DXIL Container Access
+The following interfaces are used to create/manipulate DXIL containers.
+```c++
+struct IDxcContainerBuilder : public IUnknown {
+  HRESULT Load(IDxcBlob *pDxilContainerHeader);      // Loads a container
+  HRESULT AddPart(UINT32 fourCC, IDxcBlob *pSource); // Add a part to container
+  HRESULT RemovePart(UINT32 fourCC);               // Remove part from container
+  // Builds a container of the given container builder state
+  HRESULT SerializeContainer(IDxcOperationResult **ppResult); 
+};
+
+struct IDxcAssembler : public IUnknown {
+  // Assemble dxil in ll or llvm bitcode to DXIL container.
+  HRESULT AssembleToContainer(
+    IDxcBlob *pShader,               // shader to assemble
+    IDxcOperationResult **ppResult); // output status, buffer, and errors
+};
+```
+
+#### Shader Reflection
+A DXIL container can be inspected and different parts can be accessed using
+IDxcContinerReflection. ID3D12ShaderReflection can also be retrieved using
```

Comment:
Isn't ID3D12ShaderReflection part of the API we should be looking at as well?

Reply:
Yes.  I wasn't sure how deep the rabbit hole goes, but you are correct.  This needs some attention here to cover ALL areas of functionality.  Obviously we cannot pop out that interface from a flat c-api. :)

Reply:
ID3D12ShaderReflection is implemented in DXC!  Interesting.  Guess that fits.  It is strange that the interface name is D3D12 runtime looking.

Reply:
I'll add it to this spec so it covers all supported interfaces DXC exposes.

Reply:
Added interfaces and structs hopefully in a good order to understand usage.